src/v/utils/tests/filtered_lower_bound_test.cc

```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    return res;
+}
+
+void validate(std::vector<std::optional<int>>& to_search, int searched_value) {
+    auto found_iter = filtered_lower_bound(
+      to_search.begin(),
+      to_search.end(),
+      searched_value,
+      optional_compare{},
+      optional_filter{});
+
+    auto linear_search_result = to_search.end();
+    for (auto iter = to_search.begin(); iter != to_search.end(); ++iter) {
+        if (iter->has_value() && iter->value() >= searched_value) {
+            linear_search_result = iter;
+            break;
+        }
+    }
+
+    auto found_dist = std::distance(to_search.begin(), found_iter);
+    auto expected_dist = std::distance(to_search.begin(), linear_search_result);
+
+    if (found_dist != expected_dist) {
+        vlog(
+          test_log.info,
+          "Failure detected: to_search={} searched_value={}, "
+          "expected_index={}, found_index={}",
+          print(to_search),
+          searched_value,
+          expected_dist,
+          found_dist);
+    }
+
+    BOOST_CHECK_EQUAL(found_dist, expected_dist);
+}
+
+BOOST_AUTO_TEST_CASE(filtered_lower_bound_test) {
+    std::vector<std::optional<int>> case1 = {1, 3, 5, std::nullopt, 7, 9};
+    for (auto val : {0, 1, 3, 5, 6, 7, 9, 100}) {
+        validate(case1, val);
+    }
+
+    std::vector<std::optional<int>> case2 = {
+      1, 3, 5, std::nullopt, std::nullopt, 7, 9};
+    for (auto val : {0, 1, 3, 5, 6, 7, 9, 100}) {
+        validate(case2, val);
+    }
+
+    std::vector<std::optional<int>> case3 = {
+      1, 3, 5, std::nullopt, std::nullopt, std::nullopt, 7, 9};
+    for (auto val : {0, 1, 3, 5, 6, 7, 9, 100}) {
+        validate(case3, val);
+    }
+
+    std::vector<std::optional<int>> case4 = {
+      1, std::nullopt, 3, 5, std::nullopt, std::nullopt, std::nullopt, 7, 9};
+    for (auto val : {0, 1, 3, 5, 6, 7, 9, 100}) {
+        validate(case4, val);
+    }
+
+    std::vector<std::optional<int>> case5 = {
+      std::nullopt, std::nullopt, std::nullopt, std::nullopt};
+    for (auto val : {0, 1}) {
+        validate(case5, val);
+    }
+
+    std::vector<std::optional<int>> case6 = {
+      0, std::nullopt, std::nullopt, std::nullopt, std::nullopt};
+    for (auto val : {-1, 0, 1}) {
+        validate(case6, val);
+    }
+
+    std::vector<std::optional<int>> case7 = {
+      std::nullopt, std::nullopt, std::nullopt, std::nullopt, 1};
+    for (auto val : {-1, 0, 1}) {
+        validate(case7, val);
+    }
+
+    std::vector<std::optional<int>> case8 = {
+      std::nullopt, 0, 1, std::nullopt, 2, 3, 4, std::nullopt, std::nullopt, 5};
+    for (auto val : {-1, 0, 1, 2, 3, 4, 5, 6}) {
+        validate(case8, val);
+    }
+
+    std::vector<std::optional<int>> case9 = {
+      std::nullopt,
+      0,
+      std::nullopt,
+      std::nullopt,
+      1,
+      2,
+      3,
+      std::nullopt,
+      std::nullopt,
+      4};
+    validate(case9, -1);
+}
+
+BOOST_AUTO_TEST_CASE(fuzz_test) {
```

Comment:
I'm fuzzing this a lot more aggressively on my machine. Hasn't failed yet.

Reply:
I've been fuzzing for about 24 hours now and haven't seen any failures. I'll stop now.